pkg/scheduler/controller/shoot/scheduler_control.go

```diff
@@ -450,12 +440,24 @@ func ignoreSeedDueToDNSConfiguration(seed *gardencorev1beta1.Seed, shoot *garden
 	return !gardencorev1beta1helper.ShootUsesUnmanagedDNS(shoot)
 }
 
-type executeSchedulingRequest = func(context.Context, *gardencorev1beta1.Shoot) error
+// tryScheduleShoot retries scheduling the shoot onto the given seed with exponential backoff.
+func tryScheduleShoot(ctx context.Context, c client.Client, shoot *gardencorev1beta1.Shoot, seedName string) error {
+	// need retry logic, because the controller-manager is acting on it at the same time: setting Status to Pending until scheduled
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := c.Get(ctx, client.ObjectKeyFromObject(shoot), shoot); err != nil {
+			return err
+		}
+
+		if shoot.Spec.SeedName != nil {
+			return common.NewAlreadyScheduledError(fmt.Sprintf("shoot is already scheduled to seed %s", *shoot.Spec.SeedName))
+		}
+
+		// run with optimistic locking to prevent rescheduling onto different seed
```

Comment:
Hmm, I was actually struggling a bit on what to do with this one.
While rechecking the details now, I realized that the reason for retrying the `spec.seedName` update is gone meanwhile (gcm no longer sets status to pending for unscheduled shoots). So, I got rid of `RetryOnConflict` here, simplified the handling to use optimistic locking and use the default rate limiter for the workqueue instead of custom rate limiter with baseDelay of 15s to make retries faster.

This makes the [`retrySyncPeriod` field](https://github.com/gardener/gardener/blob/278e9b4e584404e14ec1b0d214987a19f4352eb0/pkg/scheduler/apis/config/v1alpha1/types.go#L104) in the `SchedulerConfig` obsolete, so I will remove it in another PR.

Still we want to use optimistic locking here to prevent scheduling based on stale data, e.g. if the shoot was manually scheduled or it was requeued or whatever. We definitely want to make sure to not reschedule the shoot in any case.
I also played around with this a bit and it seems like the chances of running into conflicts are basically 0 and also the default workqueue works perfectly fine, I don't see any issues of switching to it like all the other controllers do.